src/main/java/com/opentable/components/filterorder/FilterOrderResolver.java

```diff
@@ -0,0 +1,134 @@
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+import javax.servlet.Filter;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.boot.web.servlet.FilterRegistrationBean;
+
+/**
+ * {@link FilterOrderResolver} resolves the order of servlet filters according to injected {@link OrderDeclaration}s.
+ *
+ * <p>
+ * The total ordering is as follows.
+ * First come the filters for which no order was declared, in original injection order.
+ * Next are the filters for which an order was declared, in the order resolved by this class.
+ * If there is a last filter, it comes last.
+ *
+ * <p>
+ * If more than one last filter is declared, or if the order declarations create a cycle, {@link #resolve} throws
+ * a runtime exception with helpful diagnostic information.
+ *
+ * <p>
+ * Inject this just before you initialize the servlet container. It will call
+ * {@link FilterRegistrationBean#setOrder}} on the beans to enforce their order.
+ *
+ * @see OrderDeclaration
+ */
+public class FilterOrderResolver {
+    private static final Logger LOG = LoggerFactory.getLogger(FilterOrderResolver.class);
+
+    private final List<FilterRegistrationBean<?>> filterRegistrationBeans;
+    private final Collection<OrderDeclaration> orderDeclarations;
+
+    @Inject
+    public FilterOrderResolver(
+            final List<FilterRegistrationBean<?>> filterRegistrationBeans,
+            final Optional<Collection<OrderDeclaration>> orderDeclarations) {
+        this.filterRegistrationBeans = filterRegistrationBeans;
+        this.orderDeclarations = orderDeclarations.orElse(Collections.emptySet());
+    }
+
+    @PostConstruct
+    void resolve() {
+        if (orderDeclarations.isEmpty()) {
+            LOG.debug("no order declarations; will not perform resolution");
+            return;
+        }
+
+        final Map<FilterRegistrationBean<?>, Integer> originalIndexes = new HashMap<>();
+        for (int i = 0; i < filterRegistrationBeans.size(); ++i) {
+            final FilterRegistrationBean<?> frb = filterRegistrationBeans.get(i);
+            originalIndexes.put(frb, i);
+        }
+        final Set<Class<? extends Filter>> last = new HashSet<>();
+        final Map<Class<? extends Filter>, Collection<Class<? extends Filter>>> dependencies = new HashMap<>();
+        orderDeclarations.forEach(od -> {
+            if (od.last) {
+                last.add(od.filter);
+                return;
+            }
+            dependencies
+                    .computeIfAbsent(od.filter, f -> new HashSet<>())
+                    .add(od.dependsOn);
+        });
+        if (last.size() != 1) {
+            throw new IllegalStateException(String.format("multiple filters declared to come last: %s", last));
```

Comment:
Might want to declare a Spring Boot FailureAnalyzer to give a user friendly error message on start-up if there's a problem like this.

Reply:
Can you elaborate on this suggestion @adamzr?  Do you not think the exception string we emit here is clear?

Reply:
See http://www.baeldung.com/spring-boot-failure-analyzer I think it'd be fine as is, it'd be a stacktrace with a decent error message. But, with a failure analyzer it would print something like:
```
***************************
APPLICATION FAILED TO START
***************************
 
Description:
 
The bean MyFilterA and MyFilterB are both declared as the last filter in the order declaration bean named mySampleOrderDeclaraton.
 
Action:
 
Change the order declarations, so that only one bean is declared as last or so that none of them are.
```

Reply:
That would indeed be pretty cool.  Although I'm a little suspicious of further coupling with boot.  About how much code and dependencies would we have to add to get this to work with this module?

Reply:
I guess I view us as already committed and coupled to Boot. I'm not sure how complicated it is to do this, but it shouldn't be hard. I think it's supposed to be an integration point. This is how Boot wants people to handle errors that would block application start-up.